java/src/main/java/com/saucelabs/simplesauce/SauceOptions.java

```diff
@@ -1,134 +1,150 @@
-        return this;
-    }
-    public SauceOptions withSafari()
-    {
-        return withMac(MacVersion.Mojave);
-    }
-
-    public SauceOptions withSafari(SafariVersion version) {
-        browserName = BrowserType.SAFARI;
-        browserVersion = version.getVersion();
-        return this;
-    }
-
-    public SauceOptions withSafari(final String version) {
-        String _version = version;
-        if (_version.isEmpty()) { _version = "latest"; }
-        browserName = BrowserType.SAFARI;
-        browserVersion = _version;
-        return this;
-    }
-
-    public SauceOptions withLinux() {
-        operatingSystem = "Linux";
-        return this;
-    }
-
-    public SauceOptions withWindows10() {
-        operatingSystem = "windows 10";
-        return this;
-    }
-    public SauceOptions withWindows8_1() {
-        operatingSystem = "Windows 8.1";
-        return this;
-    }
-    public SauceOptions withWindows8() {
-        operatingSystem = "Windows 8";
-        return this;
-    }
-
-    public SauceOptions withWindows7() {
-        operatingSystem = "Windows 7";
-        return this;
-    }
-
-    public SauceOptions withEdge() {
-        browserName = "Edge";
-        browserVersion = "18.17763";
-        return this;
-    }
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
 
-    //TODO notice the duplication below with edge.
-    //Maybe could be moved to a separate class so we can do withEdge().16_16299();
-    //Or withEdge().version(EdgeVersion.16_16299);
-    public SauceOptions withEdge18() {
-        withEdge();
-        browserVersion = "18.17763";
-        return this;
-    }
-    public SauceOptions withEdge17() {
-        withEdge();
-        browserVersion = "17.17134";
-        return this;
-    }
-    public SauceOptions withEdge16() {
-        withEdge();
-        browserVersion = "16.16299";
-        return this;
-    }
-
-    public SauceOptions withEdge15() {
-        withEdge();
-        browserVersion = "15.15063";
-        return this;
-    }
-
-    public SauceOptions withEdge14() {
-        withEdge();
-        browserVersion = "14.14393";
-        return this;
-    }
-
-    public SauceOptions withEdge13() {
-        withEdge();
-        browserVersion = "13.10586";
-        return this;
-    }
-
-    public SauceOptions withFirefox() {
-        browserName = "Firefox";
-        return this;
-    }
-
-    public SauceOptions withIE(IEVersion version) {
-        this.browserVersion = version.getVersion();
-        return this;
-    }
-
-    public SauceOptions withIE(String version) {
-        browserName = "IE";
-        browserVersion = version;
-        return this;
-    }
-
-    public SauceOptions withIE() {
-        browserName = "IE";
-        browserVersion = "latest";
-        return this;
-    }
+public class SauceOptions {
```

Comment:
Consider implementing `Capabilities` so that this can be passed directly to a `RemoteWebDriver`:

```java
var options = new SauceOptions();
var driver = new RemoteWebDriver(options);
```

Reply:
We did discuss this idea, or at least subclassing `SauceOptions` as a `MutableCapabilities`. I don't recall why we didn't do this but there was a good reason.

Reply:
My current reasons for not doing this (and I might spike out an implementation to verify):
1. SauceOptions is designed to be consumed by SauceSession (for "reasons") not RemoteWebDriver, so it would be confusing to create an instance that "shouldn't" be used in other places
2. I'd like to be strict about what fields are allowed in this class (e.g. not supporting a generic `setCapability`)
3. I don't think it would be compatible with lombok, which I'm not tied to, but it would mean hard coding all of the setter/getters or abandoning point 2.

Reply:
Fair enough. If you're never expecting to pass this to a `RemoteWebDriver`, then things are a little different.